### PR TITLE
Add stan module

### PR DESCRIPTION
**Pull request by G-Kazantzidis**

### CODE REVIEW

File: R/testfile.R
Comment (gowerc) [inline R/testfile.R:1]: Not sure why this has been added :)

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:49]: Instead of an explicitly function I think instead we should be using the "initialise" method. This is a magic method that is automatically called when users run the `StanModules()` function

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:30]: I think there needs to be logic here that can handle the scenario of the user actually providing the code directory instead of a file path something like (pseudo code):

```
function (string) {
    if  is_file(string):
        return  read_file(absolute_filename)
    else:
        return string
}
```

Comment (G-Kazantzidis) [reply R/stan_module.R:30]: I use `file.exists` to check whether the string is an existing file and otherwise I use the `system.file` function to locate the file.

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:41]: Sorry this logic doesn't make sense to me.   If the user were to provide the stan code as a string i.e.

```
parameters = "
    real beta;
    real gamma;
    real sigma;
"
```

This would error because line 34 would try and find it as as a system file ?

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:59]: Should this be `priors` to be consistent with the pluralisation we've put on the other function argument names ?

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:70]: I think its fine for this to be length 0, maybe initialise the default arguments with just `""`.  Because I can imagine it being valid to create a stan block which doesn't have one of the sections filled in, for example the HazardLink objects aren't going to need to populate the data block or the generated_quatities block ?

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:22]: You seem to have reverted to only 2 space indents here ?

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R]: Indeed, allowing the user to add the code as a text could be helpful. Although I will keep the option of the `system.file` as well. It is important because we want to locate the examples we are going to provide along with the package.

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:32]: I think this line can be deleted and just pass `file` to the `readLines` function

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:37]: Perhaps collapse into a single ` } else if {` block. At least I'm not sure the nested if is needed ?

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:48]: Does this not error because the argument is now just `file` ? Perhaps rename as `string` or something generic to avoid confusion as it can now be either a string or a filename

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R]: Looks like you have a mix of 4 and 2 spacing here

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R]: mix of 4 and 2 spacing

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R]: Looks like this is just 2 spacing

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R:66]: I added the `paste_str` function to paste the elements of a vector if the user provides a vector

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:39]: You end up calling the system.file function twice here with identical arguments, perhaps just call it once in advance and assign the value to a variable.

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:116]: It would be good to add some unit tests to show this function is working as expected

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:119]: Shouldn't this be a check that the line ends with ";"  for example what about the string `real sigma; real beta`

Comment (gowerc) [reply R/stan_module.R:119]: Also the `== TRUE` is not needed

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:123]: I'm not sure its even possible to enter the `else` block here is it ?

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R:119]: > Also the `== TRUE` is not needed

but then, if `grepl(";", vec, fixed = TRUE)` gives only `FALSE`, the if condition will be `TRUE`.

File: R/stan_module.R
Comment (gowerc) [reply R/stan_module.R:123]: ah I see when we have differences in some having semi colons and others not gotcha.

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:83]: Subjective but I'm wondering if it makes sense to move the  `if(length(...)> 1` logic inside of the `paste_str` function so that we don't have to have these repeated `if()` blocks ?

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:77]: I think we need to allow for `priors` being populated and `inits` being left blank. This currently forces you to define your own `init` even if you'd rather leave it up to Stan to do that for you

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R:61]: Okay, I am not an S4 expert either. I agree with this approach.

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R:91]: Another option would be to allow for `length` to be `>1`. And we can paste it together later.

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R:85]: The idea is that the default `inits` will be a list of either `NAs`, values or `functions`. This list will be curried on until the moment when the data are merged to the same object as the code. Then, the functions within the `inits` list will generate the necessary initial values.  For example the default list of the initial values of the os model https://github.com/Genentech/jmpost/blob/c9b971708856990559d88e31026a2402b1bf0fd9/R/inits.R#L25-L29

File: R/stan_module.R
Comment (gowerc) [reply R/stan_module.R:91]: Potentially yes we could handle this in the is.character method.  Would also make merging StanModules easier as we don't need to worry about the logic of collapsing them there.

File: R/stan_module.R
Comment (gowerc) [reply R/stan_module.R:72]: I think it would be good to assign sensible defaults here  to avoid users running into this error which might be quite annoying, perhaps something like:

```
    definition = function(
        .Object,
        <other arguments>
        priors = list(),
        inits =  rep(NA, length(priors))
    )
```

Comment (G-Kazantzidis) [reply R/stan_module.R:72]: >         inits =  rep(NA, length(priors))

The issue here is that the inits list might contain an element that is a list or a vector of length defined from the data object. Thus, the stan code will expect eg `inits$beta_os_cov` to be a vector of length 5 and it will get only one `NA`.

Comment (gowerc) [reply R/stan_module.R:72]: Fair point, maybe instead we should just support `inits` being a single `NA` value which we can then replicate across later when we reduce the code

Comment (G-Kazantzidis) [reply R/stan_module.R:72]: What about being a list of either `NA` values or functions? I do believe that there is a benefit of carrying this list of values or functions from the beginning. This way, the user can modify this list at any point.

Comment (gowerc) [reply R/stan_module.R:74]: O yes, sorry I'm not saying not support functions / list of values, what I'm trying to implement is a simple way for the user to say "I don't want to set priors for anything"  which I think just providing a single NA value (or a null or a blank list) would be quite clean.

Thinking about it I think a blank list is fine,  I  am thinking that if any parameter is missing from inits then we should assume they dont want to specify a starting value for that parameter (assuming stan can support this)

**Comment by gowerc** [comment]
I'm currently getting an error when I try to use the object using non-file code btw: 

```
> StanModule(
+     functions = "myfun(){awdawda}"
+ )
Error in file.exists(file) : invalid 'file' argument
```

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:185]: @G-Kazantzidis , sorry I thought we said in another chat that we would handle concatination in the as.character method and that we would suppoer  data/functions/etc  being > length 1.  In this case I would have expected this to just be a combining of vectors  i.e.  

```
functions =  c(x@functions, y@functions)
```

and then we have  a single place to handle  `\\n` and `;` in the as.character method?

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:171]: I think you need to export this method

File: R/stan_module.R
Comment (G-Kazantzidis) [reply R/stan_module.R:76]: I see your point. Lets allow the user to provide an empty list at this stage.

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:84]: How come you've used `validate_that` instead of `assert_that`, don't we want to throw an error if the condition isn't met ?

Comment (G-Kazantzidis) [reply R/stan_module.R:84]: Because of the case that the user defines the prior list (which is essential for the model built) but not the initial values. In that case, the lengths of the two lists are going to be different.

**Review by gowerc — APPROVED** [review]

**Pull request merged**
